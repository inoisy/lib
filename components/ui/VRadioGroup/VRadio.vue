<script>
/* eslint-disable no-empty-function,no-unused-vars,vue/require-prop-types,no-prototype-builtins,new-cap,vue/require-default-prop,no-unused-expressions */

// Components
import VLabel from '../VLabel';
import VIcon from '../VIcon';
import VInput from '../VInput';

// Mixins
import BindsAttrs from '../../../mixins/binds-attrs';
import Colorable from '../../../mixins/colorable';
import { factory as GroupableFactory } from '../../../mixins/groupable';
import Rippleable from '../../../mixins/rippleable';
import Themeable from '../../../mixins/themeable';
import Selectable, { prevent } from '../../../mixins/selectable';

// Utilities
import { getSlot } from '../../../utils/helpers';

import mixins from '../../../utils/mixins';
import { mergeListeners } from '../../../utils/mergeData';

const baseMixins = mixins(BindsAttrs, Colorable, Rippleable, GroupableFactory('radioGroup'), Themeable);

export default baseMixins.extend({
    name: 'VRadio',

    inheritAttrs: false,

    props: {
        disabled: Boolean,
        id: String,
        label: String,
        name: String,
        offIcon: {
            type: String,
            default: '$radioOff',
        },
        onIcon: {
            type: String,
            default: '$radioOn',
        },
        readonly: Boolean,
        value: {
            default: null,
        },
    },

    data() {
        return {
            isFocused: false,
        };
    },

    computed: {
        classes() {
            return {
                'v-radio--is-disabled': this.isDisabled,
                'v-radio--is-focused': this.isFocused,
                ...this.themeClasses,
                ...this.groupClasses,
            };
        },
        computedColor() {
            return Selectable.options.computed.computedColor.call(this);
        },
        computedIcon() {
            return this.isActive
                ? this.onIcon
                : this.offIcon;
        },
        computedId() {
            return VInput.options.computed.computedId.call(this);
        },
        hasLabel: VInput.options.computed.hasLabel,
        hasState() {
            return (this.radioGroup || {}).hasState;
        },
        isDisabled() {
            return this.disabled ||
                Boolean(this.radioGroup) &&
                this.radioGroup.isDisabled;
        },
        isReadonly() {
            return this.readonly ||
                Boolean(this.radioGroup) &&
                this.radioGroup.isReadonly;
        },
        computedName() {
            if (this.name || !this.radioGroup) {
                return this.name;
            }

            return this.radioGroup.name || `radio-${this.radioGroup._uid}`;
        },
        rippleState() {
            return Selectable.options.computed.rippleState.call(this);
        },
        validationState() {
            return (this.radioGroup || {}).validationState || this.computedColor;
        },
    },

    methods: {
        genInput(args) {
            // We can't actually use the mixin directly because
            // it's made for standalone components, but its
            // genInput method is exactly what we need
            return Selectable.options.methods.genInput.call(this, 'radio', args);
        },
        genLabel() {
            if (!this.hasLabel) {
                return null;
            }

            return this.$createElement(VLabel, {
                on: {
                    // Label shouldn't cause the input to focus
                    click: prevent,
                },
                attrs: {
                    for: this.computedId,
                },
                props: {
                    color: this.validationState,
                    focused: this.hasState,
                },
            }, getSlot(this, 'label') || this.label);
        },
        genRadio() {
            const { title, ...radioAttrs } = this.attrs$;

            return this.$createElement('div', {
                staticClass: 'v-input--selection-controls__input',
            }, [
                this.$createElement(VIcon, this.setTextColor(this.validationState, {
                    props: {
                        dense: this.radioGroup && this.radioGroup.dense,
                    },
                }), this.computedIcon),
                this.genInput({
                    name: this.computedName,
                    value: this.value,
                    ...radioAttrs,
                }),
                this.genRipple(this.setTextColor(this.rippleState)),
            ]);
        },
        onFocus(e) {
            this.isFocused = true;
            this.$emit('focus', e);
        },
        onBlur(e) {
            this.isFocused = false;
            this.$emit('blur', e);
        },
        onChange() {
            if (this.isDisabled || this.isReadonly || this.isActive) {
                return;
            }

            this.toggle();
        },
        onKeydown: () => {}, // Override default with noop
    },

    render(h) {
        const data = {
            staticClass: 'v-radio',
            class: this.classes,
            on: mergeListeners({
                click: this.onChange,
            }, this.listeners$),
            attrs: { title: this.attrs$.title },
        };

        return h('div', data, [
            this.genRadio(),
            this.genLabel(),
        ]);
    },
});

</script>
<style lang="scss">
    @import './_variables';


    /* stylelint-disable declaration-no-important,no-duplicate-selectors */

    @include theme(v-radio) using ($material) {
        &--is-disabled {
            label {
                color: map-deep-get($material, 'text', 'disabled');
            }

            .v-icon {
                // needed for helper override
                color: map-deep-get($material, 'selection-controls', 'disabled') !important;
            }
        }
    }

    .v-radio {
        display: flex;
        align-items: center;
        height: auto;
        outline: none;

        &--is-disabled {
            cursor: default;
            pointer-events: none;
        }
    }

    .v-input--radio-group.v-input--radio-group--row {
        .v-radio {
            margin-right: $radio-margin-right;
        }
    }
</style>
