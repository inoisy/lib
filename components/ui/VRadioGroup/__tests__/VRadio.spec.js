/* eslint-disable no-empty-function */
// Components
import VRadio from '../VRadio';

// Utilities
import {
    mount,
} from '@vue/test-utils';

describe('VRadio.js', () => {
    let mountFunction;

    beforeEach(() => {
        mountFunction = (options = {}) => mount(VRadio, options);
    });

    it('should render role and aria-checked attributes on input group', async () => {
        const wrapper = mountFunction({
            data: () => ({
                isActive: false,
            }),
            provide: {
                radio: {
                    name: 'name',
                    isMandatory: false,
                },
            },
        });

        let inputGroup = wrapper.find('input');
        expect(inputGroup.element.getAttribute('role')).toBe('radio');
        expect(inputGroup.element.getAttribute('aria-checked')).toBe('false');

        await wrapper.setData({ isActive: true });
        inputGroup = wrapper.find('input');
        expect(inputGroup.element.getAttribute('aria-checked')).toBe('true');
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should not render aria-label attribute with no label value on input group', () => {
        const wrapper = mountFunction({
            propsData: {
                label: null,
            },
            provide: {
                radio: {
                    name: 'name',
                    isMandatory: false,
                },
            },
        });

        const inputGroup = wrapper.find('input');
        expect(inputGroup.element.getAttribute('aria-label')).toBeFalsy();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render proper input name', () => {
        const wrapper = mountFunction({
            provide: {
                radioGroup: {
                    name: 'name',
                    register: () => {},
                    unregister: () => {},
                },
            },
        });

        const input = wrapper.find('input');
        expect(input.element.getAttribute('name')).toBe('name');
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should toggle on keypress', () => {
        const wrapper = mountFunction();

        const change = jest.fn();
        wrapper.vm.$on('change', change);

        const input = wrapper.find('input');

        input.trigger('change');
        expect(change).toHaveBeenCalledTimes(1);

        input.trigger('keydown.tab');
        expect(change).toHaveBeenCalledTimes(1);
    });

    it('should use custom icons', async () => {
        const wrapper = mountFunction({
            propsData: {
                onIcon: 'foo',
                offIcon: 'bar',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();

        await wrapper.setData({ isActive: true });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should check/uncheck the internal input', async () => {
        const wrapper = mountFunction();

        expect(wrapper.vm.$refs.input.checked).toBeFalsy();

        await wrapper.setData({ isActive: true });

        expect(wrapper.vm.$refs.input.checked).toBeTruthy();

        await wrapper.setData({ isActive: false });

        expect(wrapper.vm.$refs.input.checked).toBeFalsy();
    });

    it('should set focused state', async () => {
        const wrapper = mountFunction();

        expect(wrapper.vm.isFocused).toBe(false);

        const input = wrapper.find('input');

        input.trigger('focus');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.isFocused).toBe(true);

        input.trigger('blur');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.isFocused).toBe(false);
    });

    it('should be render colored radio', () => {
        const wrapper = mountFunction({
            propsData: { color: 'yellow' },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});


// // Components
// import VRadio from '../VRadio';

// // Utilities
// import {
//     mount,
// } from '@vue/test-utils';

// describe('VRadio.js', () => {
//     let mountFunction;

//     beforeEach(() => {
//         mountFunction = (options = {}) => mount(VRadio, options);
//     });

//     it('should render role and aria-checked attributes on input group', async () => {
//         const wrapper = mountFunction({
//             data: () => ({
//                 isActive: false,
//             }),
//             provide: {
//                 radio: {
//                     name: 'name',
//                     isMandatory: false,
//                 },
//             },
//         });

//         let inputGroup = wrapper.find('input');
//         expect(inputGroup.element.getAttribute('role')).toBe('radio');
//         expect(inputGroup.element.getAttribute('aria-checked')).toBe('false');

//         await wrapper.setData({ isActive: true });
//         inputGroup = wrapper.find('input');
//         expect(inputGroup.element.getAttribute('aria-checked')).toBe('true');
//         expect(wrapper.html()).toMatchSnapshot();
//     });

//     it('should not render aria-label attribute with no label value on input group', () => {
//         const wrapper = mountFunction({
//             propsData: {
//                 label: null,
//             },
//             provide: {
//                 radio: {
//                     name: 'name',
//                     isMandatory: false,
//                 },
//             },
//         });

//         const inputGroup = wrapper.find('input');
//         expect(inputGroup.element.getAttribute('aria-label')).toBeFalsy();
//         expect(wrapper.html()).toMatchSnapshot();
//     });

//     it('should render proper input name', () => {
//         const wrapper = mountFunction({
//             provide: {
//                 radioGroup: {
//                     name: 'name',
//                     register: () => {},
//                     unregister: () => {},
//                 },
//             },
//         });

//         const input = wrapper.find('input');
//         expect(input.element.getAttribute('name')).toBe('name');
//         expect(wrapper.html()).toMatchSnapshot();
//     });

//     it('should toggle on keypress', () => {
//         const wrapper = mountFunction();

//         const change = jest.fn();
//         wrapper.vm.$on('change', change);

//         const input = wrapper.find('input');

//         input.trigger('change');
//         expect(change).toHaveBeenCalledTimes(1);

//         input.trigger('keydown.tab');
//         expect(change).toHaveBeenCalledTimes(1);
//     });

//     it('should use custom icons', async () => {
//         const wrapper = mountFunction({
//             propsData: {
//                 onIcon: 'foo',
//                 offIcon: 'bar',
//             },
//         });

//         expect(wrapper.html()).toMatchSnapshot();

//         await wrapper.setData({ isActive: true });

//         expect(wrapper.html()).toMatchSnapshot();
//     });

//     it('should check/uncheck the internal input', async () => {
//         const wrapper = mountFunction();

//         expect(wrapper.vm.$refs.input.checked).toBeFalsy();

//         await wrapper.setData({ isActive: true });

//         expect(wrapper.vm.$refs.input.checked).toBeTruthy();

//         await wrapper.setData({ isActive: false });

//         expect(wrapper.vm.$refs.input.checked).toBeFalsy();
//     });

//     it('should set focused state', async () => {
//         const wrapper = mountFunction();
//         await wrapper.vm.$nextTick();

//         expect(wrapper.vm.isFocused).toBe(false);

//         const input = wrapper.find('input');

//         input.trigger('focus');
//         await wrapper.vm.$nextTick();

//         expect(wrapper.vm.isFocused).toBe(true);

//         input.trigger('blur');
//         await wrapper.vm.$nextTick();

//         expect(wrapper.vm.isFocused).toBe(false);
//     });

//     it('should be render colored radio', () => {
//         const wrapper = mountFunction({
//             propsData: { color: 'yellow' },
//         });

//         expect(wrapper.html()).toMatchSnapshot();
//     });
// });
