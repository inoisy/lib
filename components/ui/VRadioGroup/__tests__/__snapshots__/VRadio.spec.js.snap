// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VRadio.js should be render colored radio 1`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light"
    >
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           role="img"
           aria-hidden="true"
           class="v-icon__svg"
      >
        <path d="M12,20C7.58,20 4,16.42 4,12C4,7.58 7.58,4 12,4C16.42,4 20,7.58 20,12C20,16.42 16.42,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2Z">
        </path>
      </svg>
    </span>
    <input aria-checked="false"
           id="input-22"
           role="radio"
           type="radio"
           value
    >
    <div class="v-input--selection-controls__ripple">
    </div>
  </div>
</div>
`;

exports[`VRadio.js should not render aria-label attribute with no label value on input group 1`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light"
    >
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           role="img"
           aria-hidden="true"
           class="v-icon__svg"
      >
        <path d="M12,20C7.58,20 4,16.42 4,12C4,7.58 7.58,4 12,4C16.42,4 20,7.58 20,12C20,16.42 16.42,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2Z">
        </path>
      </svg>
    </span>
    <input aria-checked="false"
           id="input-4"
           role="radio"
           type="radio"
           value
    >
    <div class="v-input--selection-controls__ripple">
    </div>
  </div>
</div>
`;

exports[`VRadio.js should render proper input name 1`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light"
    >
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           role="img"
           aria-hidden="true"
           class="v-icon__svg"
      >
        <path d="M12,20C7.58,20 4,16.42 4,12C4,7.58 7.58,4 12,4C16.42,4 20,7.58 20,12C20,16.42 16.42,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2Z">
        </path>
      </svg>
    </span>
    <input aria-checked="false"
           id="input-7"
           role="radio"
           type="radio"
           name="name"
           value
    >
    <div class="v-input--selection-controls__ripple">
    </div>
  </div>
</div>
`;

exports[`VRadio.js should render role and aria-checked attributes on input group 1`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light primary--text"
    >
      <svg xmlns="http://www.w3.org/2000/svg"
           viewbox="0 0 24 24"
           role="img"
           aria-hidden="true"
           class="v-icon__svg"
      >
        <path d="M12,20C7.58,20 4,16.42 4,12C4,7.58 7.58,4 12,4C16.42,4 20,7.58 20,12C20,16.42 16.42,20 12,20M12,2C6.48,2 2,6.48 2,12C2,17.52 6.48,22 12,22C17.52,22 22,17.52 22,12C22,6.48 17.52,2 12,2M12,7C9.24,7 7,9.24 7,12C7,14.76 9.24,17 12,17C14.76,17 17,14.76 17,12C17,9.24 14.76,7 12,7Z">
        </path>
      </svg>
    </span>
    <input aria-checked="true"
           id="input-1"
           role="radio"
           type="radio"
           value
    >
    <div class="v-input--selection-controls__ripple primary--text">
    </div>
  </div>
</div>
`;

exports[`VRadio.js should use custom icons 1`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light"
    >
      <svg-icon class="v-icon__component theme--light">
      </svg-icon>
    </span>
    <input aria-checked="false"
           id="input-13"
           role="radio"
           type="radio"
           value
    >
    <div class="v-input--selection-controls__ripple">
    </div>
  </div>
</div>
`;

exports[`VRadio.js should use custom icons 2`] = `
<div class="v-radio theme--light">
  <div class="v-input--selection-controls__input">
    <span aria-hidden="true"
          class="v-icon notranslate theme--light primary--text"
    >
      <svg-icon class="v-icon__component theme--light">
      </svg-icon>
    </span>
    <input aria-checked="true"
           id="input-13"
           role="radio"
           type="radio"
           value
    >
    <div class="v-input--selection-controls__ripple primary--text">
    </div>
  </div>
</div>
`;
